### ISSUE TITLE
apollo binary not found while generating the scaffold

### ISSUE THREAD

**Issue by vkasraj** [mobxjs/mst-gql#130]
### Package.json
```json
{
    "main": "./src/index.tsx",
    "scripts": {
        "start": "expo start",
        "android": "expo start --android",
        "ios": "expo start --ios",
        "web": "expo start --web",
        "eject": "expo eject",
        "build:apk": "expo build:android -t apk",
        "generate": "mst-gql"
    },
    "dependencies": {
        "@react-native-community/masked-view": "^0.1.1",
        "@react-navigation/core": "^5.0.0-alpha.16",
        "@react-navigation/native": "next",
        "@react-navigation/stack": "^5.0.0-alpha.27",
        "expo": "^35.0.0",
        "formik": "^1.5.8",
        "graphql-request": "^1.8.2",
        "mobx": "^5.14.2",
        "mobx-react-lite": "^1.5.0",
        "mobx-state-tree": "^3.14.1",
        "mst-gql": "^0.6.0",
        "react": "16.8.3",
        "react-dom": "16.8.3",
        "react-native": "https://github.com/expo/react-native/archive/sdk-35.0.0.tar.gz",
        "react-native-gesture-handler": "~1.3.0",
        "react-native-paper": "^3.0.0-alpha.7",
        "react-native-reanimated": "~1.2.0",
        "react-native-safe-area-context": "~0.3.6",
        "react-native-screens": "~1.0.0-alpha.23",
        "react-native-web": "^0.11.7"
    },
    "devDependencies": {
        "@types/react": "^16.8.23",
        "@types/react-native": "^0.57.65",
        "babel-preset-expo": "^7.1.0",
        "typescript": "^3.6.3"
    }
}
```

### mst-gql.config.js
```js
module.exports = {
    format: "ts",
    input: "http://localhost:8000",
    outDir: "src/models"
};
```

### Error
```bash
$> npm run generate

> @ generate /home/hello/Documents/code/keeper/native
> mst-gql

mst-gql-scaffold.js --format=ts --outDir=/home/hello/Documents/code/keeper/native/src/models http://localhost:8000
/bin/sh: /home/hello/Documents/code/keeper/native/node_modules/mst-gql/generator/../node_modules/.bin/apollo: No such file or directory
child_process.js:660
    throw err;
    ^

Error: Command failed: /home/hello/Documents/code/keeper/native/node_modules/mst-gql/generator/../node_modules/.bin/apollo schema:download --e
ndpoint=http://localhost:8000 tmp_schema.json
/bin/sh: /home/hello/Documents/code/keeper/native/node_modules/mst-gql/generator/../node_modules/.bin/apollo: No such file or directory

    at checkExecSyncError (child_process.js:621:11)
    at Object.execSync (child_process.js:657:15)
    at main (/home/hello/Documents/code/keeper/native/node_modules/mst-gql/generator/mst-gql-scaffold.js:68:19)
    at Object.<anonymous> (/home/hello/Documents/code/keeper/native/node_modules/mst-gql/generator/mst-gql-scaffold.js:111:1)
    at Module._compile (internal/modules/cjs/loader.js:945:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:962:10)
    at Module.load (internal/modules/cjs/loader.js:798:32)
    at Function.Module._load (internal/modules/cjs/loader.js:711:12)
    at Function.Module.runMain (internal/modules/cjs/loader.js:1014:10)
    at internal/main/run_main_module.js:17:11 {
  status: 127,
  signal: null,
  output: [
    null,
    <Buffer >,
    <Buffer 2f 62 69 6e 2f 73 68 3a 20 2f 68 6f 6d 65 2f 68 65 6c 6c 6f 2f 44 6f 63 75 6d 65 6e 74 73 2f 63 6f 64 65 2f 6b 65 65 70 65 72 2f 6e 61 74 69 76 65 2f ... 86 more bytes>
  ],
  pid: 6770,
  stdout: <Buffer >,
  stderr: <Buffer 2f 62 69 6e 2f 73 68 3a 20 2f 68 6f 6d 65 2f 68 65 6c 6c 6f 2f 44 6f 63 75 6d 65 6e 74 73 2f 63 6f 64 65 2f 6b 65 65 70 65 72 2f 6e 61 74 69 76 65 2f ... 86 more bytes>
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ generate: `mst-gql`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @ generate script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/hello/.npm/_logs/2019-10-20T08_41_02_955Z-debug.log
```

**Comment by odjhey** [comment]
hellow, can you show these files/folders

`/home/hello/Documents/code/keeper/native/node_modules/.bin/mst-gql` and
`/home/hello/Documents/code/keeper/native/node_modules/mst-gql/.bin/`

**Comment by vkasraj** [comment]
This folder is missing:
`/home/hello/Documents/code/keeper/native/node_modules/mst-gql/.bin/`

`/home/hello/Documents/code/keeper/native/node_modules/.bin/mst-gql`

```js
#!/usr/bin/env node
const arg = require("arg")
const path = require("path")
const fs = require("fs")
const child_process = require("child_process")
const graphql = require("graphql")

const { getConfig, mergeConfigs } = require("./config")
const { generate, writeFiles, logUnexpectedFiles } = require("./generate")

const definition = {
  "--format": String,
  "--outDir": String,
  "--roots": String,
  "--excludes": String,
  "--modelsOnly": Boolean,
  "--force": Boolean,
  "--noReact": Boolean,
  "--separate": Boolean
}

function main() {
  let args, config

  try {
    args = arg(definition)
    config = getConfig()
  } catch (e) {
    console.error(
      "Example usage: mstgql-scaffold --format=js|ts --outDir=src/models graphql-schema.json\n, valid options: " +
        Object.keys(definition).join(", ")
    )
    throw e
  }

  const {
    format,
    outDir,
    input,
    roots,
    excludes,
    modelsOnly,
    forceAll,
    noReact
  } = mergeConfigs(args, config)
  const separate = !!args["--separate"]

  console.log(
    path.basename(__filename) +
      " --format=" +
      format +
      " --outDir=" +
      outDir +
      " " +
      input
  )
  if (!/^(ts|js|mjs)$/.test(format)) {
    console.error("Invalid format parameter, expected 'js' or 'ts' or 'mjs'")
    process.exit(1)
  }
  if (!fs.existsSync(outDir)) {
    fs.mkdirSync(outDir, { recursive: true })
  }

  let json
  if (input.startsWith("http:") || input.startsWith("https:")) {
    const tmpFile = "tmp_schema.json"
    child_process.execSync(
      `${__dirname}/../node_modules/.bin/apollo schema:download --endpoint=${input} ${tmpFile}`
    )
    json = JSON.parse(fs.readFileSync(tmpFile, "utf8"))
    fs.unlinkSync(tmpFile)
  } else if (input.endsWith(".json")) {
    json = JSON.parse(fs.readFileSync(input, "utf8"))
  } else if (input.endsWith(".graphql")) {
    // Tnx https://blog.apollographql.com/three-ways-to-represent-your-graphql-schema-a41f4175100d!
    const text = fs.readFileSync(input, "utf8")
    const schema = graphql.buildSchema(text)
    const res = graphql.graphqlSync(schema, graphql.introspectionQuery)
    if (res.data) json = res.data
    else {
      console.error("graphql parse error:\n\n" + JSON.stringify(res, null, 2))
      process.exit(1)
    }
  } else {
    console.error(
      "Expected json, graphql or url as input parameter, got: " + input
    )
    process.exit(1)
  }

  console.log(
    "Detected types: \n" +
      json.__schema.types.map(t => `  - [${t.kind}] ${t.name}`).join("\n")
  )

  // console.log(JSON.stringify(json, null, 2))
  const files = generate(
    json.__schema.types,
    format,
    roots,
    excludes,
    new Date().toUTCString(),
    modelsOnly,
    noReact
  )
  writeFiles(outDir, files, format, forceAll, true, separate)
  logUnexpectedFiles(outDir, files)
}

main()
```

**Comment by odjhey** [comment]
Hmm. `.bin` folder should be there upon install.

can you try doing `npm i` or `yarn` inside `/home/hello/Documents/code/keeper/native/node_modules/mst-gql/` for the mean time. 
See if `.bin` gets built.

**Comment by vkasraj** [comment]
I did `npm i` multiple times. But no luck.

node: v12.13.0 LTS
npm: v6.12.0

I will try to reinstall the dependencies, will tell you after that If it is solved or not.

**Comment by odjhey** [comment]
this should not happen but try doing `npm i` **inside** `/home/hello/Documents/code/keeper/native/node_modules/mst-gql/`

> I did `npm i` multiple times. But no luck.

**Comment by vkasraj** [comment]
I did that but sadly not `.bin` folder as you said. But running the same script from above generating this error.

```bash
> @ mst /home/hello/Documents/code/keeper/native
> mst-gql --format ts http://localhost:8000

mst-gql-scaffold.js --format=ts --outDir=/home/hello/Documents/code/keeper/native/src/models http://localhost:8000
Error: Cannot use GraphQLSchema "[object GraphQLSchema]" from another module or realm.
Ensure that there is only one instance of "graphql" in the node_modules
directory. If different versions of "graphql" are the dependencies of other
relied on modules, use "resolutions" to ensure only one version is installed.
https://yarnpkg.com/en/docs/selective-version-resolutions
Duplicate "graphql" modules cannot be used at the same time since different
versions may have different capabilities and behavior. The data from one
version used in the function from another could produce confusing and
spurious results.
    at instanceOf (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/jsutils/instanceOf.js:28:13)
    at isSchema (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/type/schema.js:36:34)
    at assertSchema (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/type/schema.js:40:8)
    at validateSchema (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/type/validate.js:44:28)
    at assertValidSchema (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/type/validate.js:68:16)
    at assertValidExecutionArguments (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/execution/execute.js:136:35)
    at executeImpl (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/execution/execute.js:86:3)
    at execute (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/execution/execute.js:64:63)
    at Object.introspectionFromSchema (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/utilities/introspectionFromSchema.js:31:37)
    at Task.task (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/lib/commands/service/download.js:21:82)
child_process.js:660
    throw err;
    ^

Error: Command failed: /home/hello/Documents/code/keeper/native/node_modules/mst-gql/generator/../node_modules/.bin/apollo schema:download --endpoint=http://localhost:8000 tmp_schema.json
Error: Cannot use GraphQLSchema "[object GraphQLSchema]" from another module or realm.
Ensure that there is only one instance of "graphql" in the node_modules
directory. If different versions of "graphql" are the dependencies of other
relied on modules, use "resolutions" to ensure only one version is installed.
https://yarnpkg.com/en/docs/selective-version-resolutions
Duplicate "graphql" modules cannot be used at the same time since different
versions may have different capabilities and behavior. The data from one
version used in the function from another could produce confusing and
spurious results.
    at instanceOf (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/jsutils/instanceOf.js:28:13)
    at isSchema (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/type/schema.js:36:34)
    at assertSchema (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/type/schema.js:40:8)
    at validateSchema (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/type/validate.js:44:28)
    at assertValidSchema (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/type/validate.js:68:16)
    at assertValidExecutionArguments (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/execution/execute.js:136:35)
    at executeImpl (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/execution/execute.js:86:3)
    at execute (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/execution/execute.js:64:63)
    at Object.introspectionFromSchema (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/node_modules/graphql/utilities/introspectionFromSchema.js:31:37)
    at Task.task (~/Documents/code/keeper/native/node_modules/mst-gql/node_modules/apollo/lib/commands/service/download.js:21:82)

    at checkExecSyncError (child_process.js:621:11)
    at Object.execSync (child_process.js:657:15)
    at main (/home/hello/Documents/code/keeper/native/node_modules/mst-gql/generator/mst-gql-scaffold.js:68:19)
    at Object.<anonymous> (/home/hello/Documents/code/keeper/native/node_modules/mst-gql/generator/mst-gql-scaffold.js:111:1)
    at Module._compile (internal/modules/cjs/loader.js:956:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:973:10)
    at Module.load (internal/modules/cjs/loader.js:812:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Function.Module.runMain (internal/modules/cjs/loader.js:1025:10)
    at internal/main/run_main_module.js:17:11 {
  status: 2,
  signal: null,
  output: [
    null,
    <Buffer 4c 6f 61 64 69 6e 67 20 41 70 6f 6c 6c 6f 20 50 72 6f 6a 65 63 74 20 5b 73 74 61 72 74 65 64 5d 0a 4c 6f 61 64 69 6e 67 20 41 70 6f 6c 6c 6f 20 50 72 ... 711 more bytes>,
    <Buffer 45 72 72 6f 72 3a 20 43 61 6e 6e 6f 74 20 75 73 65 20 47 72 61 70 68 51 4c 53 63 68 65 6d 61 20 22 5b 6f 62 6a 65 63 74 20 47 72 61 70 68 51 4c 53 63 ... 1992 more bytes>
  ],
  pid: 18757,
  stdout: <Buffer 4c 6f 61 64 69 6e 67 20 41 70 6f 6c 6c 6f 20 50 72 6f 6a 65 63 74 20 5b 73 74 61 72 74 65 64 5d 0a 4c 6f 61 64 69 6e 67 20 41 70 6f 6c 6c 6f 20 50 72 ... 711 more bytes>,
  stderr: <Buffer 45 72 72 6f 72 3a 20 43 61 6e 6e 6f 74 20 75 73 65 20 47 72 61 70 68 51 4c 53 63 68 65 6d 61 20 22 5b 6f 62 6a 65 63 74 20 47 72 61 70 68 51 4c 53 63 ... 1992 more bytes>
}
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @ mst: `mst-gql --format ts http://localhost:8000`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the @ mst script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/hello/.npm/_logs/2019-10-23T07_05_53_706Z-debug.log
```

**Comment by odjhey** [comment]
Hello, i am not sure why this is happening.
We initially encountered this a few months back and did a manual patch as it only happens to some machines. Then forgot about it cause suddenly it no longer throw the error on new builds. maybe a search on why `npm` `node` `.bin` folder is not getting generated on some modules be a good way to start.

below is the manual temp fix/patch we did to move along, basically, we tried to resolve the `.bin/apollo` in the main `node_modules` to get it work.

`node_modules/mst-gql/generator/mst-gql-scaffold.js`
```js
function main() {
  //...
  let json
  if (input.startsWith("http:") || input.startsWith("https:")) {
    const tmpFile = "tmp_schema.json"
    //resolve to apollo 2 dirs higher, temp fix
    const xapolloPath = path.resolve(`${__dirname}/../../../node_modules/.bin/apollo`)
    child_process.execSync(
      `${xapolloPath} schema:download --endpoint=${input} ${tmpFile}`
    )
    json = JSON.parse(fs.readFileSync(tmpFile, "utf8")) 
  //...
```

> ```shell
> mst-gql-scaffold.js --format=ts --outDir=/home/hello/Documents/code/keeper/native/src/models http://localhost:8000
> /bin/sh: /home/hello/Documents/code/keeper/native/node_modules/mst-gql/generator/../node_modules/.bin/apollo: No such file or directory
> child_process.js:660
>     throw err;
> ```

**Comment by vkasraj** [comment]
I searched as you said. Most of them saying there is some issue with `npm`. But `node` and `npm` are working fine for me. 

BTW I am using https://github.com/tj/n for version management, could be this causing the issue.

FYI, I just set up my project on a different machine, the same issue came up. I literally don't know what is causing this issue and Its really holding me up to use this package.

**Comment by odjhey** [comment]
Hmm. can you share a repo that has this problem. Let me try on my machine.

> I searched as you said. Most of them saying there is some issue with `npm`. But `node` and `npm` are working fine for me.
> 
> BTW I am using https://github.com/tj/n for version management, could be this causing the issue.
> 
> FYI, I just set up my project on a different machine, the same issue came up. I literally don't know what is causing this issue and Its really holding me up to use this package.

**Comment by vkasraj** [comment]
Yeah sure. This is the project which i am working on.

https://github.com/vkasraj/keeper

**Comment by odjhey** [comment]
this repo worked for me
npm: v6.12.0
node: v12.13.0

your best bet is to try without `n`, not really sure why.
i tried using `nvm`, also worked.

i think this is no longer an mst-gql issue, we should close for now.

> Yeah sure. This is the project which i am working on.
> 
> https://github.com/vkasraj/keeper

**Comment by vkasraj** [comment]
I will try with `nvm`. Hope it should work.

If it works with `nvm` then it should be `n` issue. I will raise an issue there.

**Comment by jonlambert** [comment]
I just ran into this issue on a fresh project. There is no `<project>/node_modules/mst-gql/.bin` directory present after running:

```
npm add mst-gql graphql-request
```

Or running: 
```
cd node_modules/mst-gql && npm i
```

Can this issue be reopened please?

**Comment by numToStr** [comment]
You can try using `yarn`. 

Earlier I was using `npm` when I faced this issue. May be the issue is with npm.

**Comment by RafaelVidaurre** [comment]
I'm having this issue with `Yarn 2`

**Comment by smokku** [comment]
I also encountered this issue.

In my project, there is no `node_modules/mst-gql/.bin` directory. But there is `node_modules/mst-gql/node_modules/.bin` with `apollo`  scripts.

I've been able to work around this by:
- adding `"apollo": "^2.33.4",` dependency to `package.json`
- editing `node_modules\mst-gql\generator\mst-gql-scaffold.js` to run `const command = `apollo client:download-schema ...` relying on $PATH
